### PR TITLE
automatic lazy bundle support

### PR DESCRIPTION
**Pull request by garlicnation**
Add the ability for vulcanize to determine bundles based on lazy imports detected through <link rel="lazy-import" tags.

### CODE REVIEW

File: src/bundler.ts [hunk 1]
```diff
@@ -262,15 +260,15 @@ class Bundler {
...
     const resolvedUrl: string = urlLib.resolve(docUrl, rawUrl);
-    const bundleUrl = manifest.bundleUrlForFile.get(resolvedUrl);
+    const bundleUrl = this.resolveBundleUrl(resolvedUrl, bundle, manifest);
     if (!bundleUrl) {
       if (reachedImports.has(resolvedUrl)) {
         dom5.remove(htmlImport);
         return;
       } else {
         reachedImports.add(resolvedUrl);
+        return;
```
Comment (justinfagnani) [inline src/bundler.ts]: I kind of like the return in the original position, and remove the one above

File: src/bundler.ts [hunk 2]
```diff
@@ -467,46 +465,26 @@ class Bundler {
...
+    }
+    const index = await buildDepsIndex(entrypoints, this.analyzer);
+    const basicBundles = generateBundles(index.entrypointToDeps);
+    const bundlesAfterStrategy = strategy(basicBundles);
+    const manifest = new BundleManifest(bundlesAfterStrategy, mapper);
+    for (const bundleEntry of manifest.bundles) {
+      const bundleUrl = bundleEntry[0];
+      const bundle = {url: bundleUrl, bundle: bundleEntry[1]};
+      const bundledAst =
+          await this._bundleDocument(bundle, manifest, bundle.bundle.files);
```
Comment (justinfagnani) [inline src/bundler.ts]: `bundle.bundle`? That's unfortunate. Are there more specific names that would help give meaning to that?

File: src/bundler.ts [hunk 3]
```diff
@@ -126,26 +126,24 @@ class Bundler {
...
       url: string,
       bundle: AssignedBundle,
-      manifest: BundleManifest): boolean|string {
+      manifest: BundleManifest): string|false {
     const targetBundle = manifest.getBundleForFile(url);
     if (!targetBundle || !targetBundle.url) {
       return false;
     }
+    for (const exclude of this._excludes) {
+      if (url.search(exclude) > -1) {
```
Comment (justinfagnani) [inline src/bundler.ts]: `String.prototype.search` takes a RegExp, and will convert a string into a RegExp via `new RexExp`. That could lead to some funky behavior with some strings.

**Comment by garlicnation** [comment]
It works because most of the support was added in the analyzer, and lazy-imports is simply treated as a special case where entrypoints can be inferred.

**Comment by stramel** [comment]
So what do we need to do to get this in?

**Comment by justinfagnani** [comment]
@stramel now that the v3 branch is merged into master, we may want to take a fresh look. I'm going to retarget this PR to master so we can preserve it and delete the v3 branch. Then @usergenic will take a look, maybe this week.

**Comment by stramel** [comment]
Great! Thanks for the update @justinfagnani!

**Comment by usergenic** [comment]
I am looking at this PR and after resolving conflicts locally, it doesn't appear to have any actual changes in it.  The `deps-index.ts` code already supports the `lazy-html-import` type and the bundle url resolution w/ excludes is already supported in another location.  There might be something subtle I'm missing, but I'm going to take a different tack and create a separate PR with tests to demonstrate handling of `<link rel="lazy-import"...>` since that's what's missing.

**Comment by usergenic** [comment]
@stramel and @justinfagnani https://github.com/Polymer/polymer-bundler/pull/487 supercedes this now

**Pull request closed without merging**